Merge sort

oleksandr-maksymikhin opened the pull request:
This pull request includes implementation of Merge sort algorithm:

Implementation of merge sort algorithm is in solutions/merge_sort.py and solutions/merge.py
Unit tests are in solutions/tests/ test_merge_sort.py
Passed automation checks (pydoc, doctest, unittest)
Please review the solution and let me know if any corrections are needed.

oleksandr-maksymikhin committed:
Created project structure (merge_sort.py, merge.py, test_merge_sort.py)

Created project structure (merge_sort.py, merge.py, test_merge_sort.py). No logic added into the files: output = input

oleksandr-maksymikhin committed:
Basic functionality of sorting with type int is working. Tests are passed

Basic functionality of sorting with type int is working. Tests are passed

oleksandr-maksymikhin committed:
Extended solution to work with any data type (int, float, char, string, bool). Test coverage

Extended solution to work with any data type (int, float, char, string, bool). Test coverage.

oleksandr-maksymikhin committed:
Added protection from side effects with test coverage

Added protection from side effects with test coverage

oleksandr-maksymikhin committed:
Completed solution. Limited functionality to process data types (int, float, string, bool)

emrebiyik commented on solutions/merge.py:
```diff
@@ -0,0 +1,86 @@
+"""
+
+
+def merge(left: list[any], right: list[any]) -> list[any]:
```
Consider improving the type annotations in the merge function signature. Currently, list[any] is used, which is very generic and doesn’t enforce the homogeneity of the input lists (left and right). You can replace list[any] with TypeVar to ensure that:
	•	Both input lists must contain elements of the same type.
	•	The return list will also contain elements of that same type.

oleksandr-maksymikhin replied:
TypeVar was added with limiting of input types to int, float, str, bool.

emrebiyik commented on solutions/merge.py:
```diff
@@ -0,0 +1,86 @@
+    [1, 2, 5, 6]
+    >>> merge([2, 5, 100500], [1, 9])
+    [1, 2, 5, 9, 100500]
+    """
```
The module-level and function-level docstrings are detailed and include helpful examples. No further changes are required here.

emrebiyik commented on solutions/merge_sort.py:
```diff
@@ -0,0 +1,74 @@
+from solutions.merge import merge
+
+
+def merge_sort(input_collection: list[any]) -> list[any]:
```
•	Replace list[any] with more specific type annotations to improve clarity and robustness.
	•	For example, use List[Union[int, float, str, bool]] or introduce a TypeVar for better flexibility.
	•	This ensures the function’s expectations about the input and output types are explicitly clear and align with the supported data types.

oleksandr-maksymikhin replied:
TypeVar was added with limiting of input types to int, float, str, bool.

emrebiyik commented on solutions/merge_sort.py:
```diff
@@ -0,0 +1,74 @@
+    [1, 2, 3]
+    >>> merge_sort([3, 2, 100500, 1])
+    [1, 2, 3, 100500]
+    """
```
The existing module-level docstring is quite descriptive, but consider including a usage example or clarifying assumptions about input data (e.g., collections must be homogeneous):
"""
    >>> merge_sort([3, 1, 4, 1, 5, 9])
    [1, 1, 3, 4, 5, 9]
    >>> merge_sort(["banana", "apple", "cherry"])
    ["apple", "banana", "cherry"]
"""

oleksandr-maksymikhin replied:
Extended examples by using int and str data types.

emrebiyik replied:
> Extended examples by using int and str data types.

Thank you for incorporating the usage examples with int and str data types! The updated docstring now provides a clearer demonstration of how the merge_sort function can be used, which is very helpful for both users and developers.

emrebiyik commented on solutions/tests/test_merge_sort.py:
```diff
@@ -0,0 +1,96 @@
+
+
+class TestMergeSort(unittest.TestCase):
+    """Test the merge_sort function."""
```
•	While the class has a docstring ("""Test the merge_sort function."""), it is very general.
	•	Consider expanding the docstring to provide a more detailed explanation of the test class’s purpose and the scenarios it covers.

oleksandr-maksymikhin replied:
Extended docstring by functions behavior description.

emrebiyik replied:
> Extended docstring by functions behavior description.

good for extending the docstring with a description of the function’s behavior. The additional detail provides more context and improves the overall clarity of the test class’s purpose.

emrebiyik commented on solutions/tests/test_merge_sort.py:
```diff
@@ -0,0 +1,96 @@
+class TestMergeSort(unittest.TestCase):
+    """Test the merge_sort function."""
+
+    def test_1_one_int_element_list(self):
```
•	Some test method names are not sufficiently descriptive (e.g., test_1_one_int_element_list). Using a more meaningful name like test_single_element_list would make the test’s purpose clearer.
	•	Avoid using numerical prefixes in test names, as they do not provide additional context.

oleksandr-maksymikhin replied:
Renaimed tests.

emrebiyik commented on solutions/tests/test_merge_sort.py:
```diff
@@ -0,0 +1,96 @@
+    """Test the merge_sort function."""
+
+    def test_1_one_int_element_list(self):
+        """It should return [1] for input [1]"""
```
•	Each test method should include a brief but descriptive docstring to explain:

```
def test_single_element_list(self):
    """
    Test that the merge_sort function correctly handles a single-element list.

    A list with only one element should return itself as it is already sorted.
    Input: [1]
    Expected Output: [1]
    """
    actual = merge_sort([1])
    expected = [1]
    self.assertEqual(actual, expected)
```

emrebiyik replied:
Suggestion:
	1.	Combine merge and merge_sort in a Single File:
	•	These two functions are inherently interdependent. The merge function is a core helper function that works in tandem with merge_sort.
	•	Combining them into one file will improve maintainability and readability, as developers can see the full context of how these two functions interact without needing to navigate across multiple files.
	•	It aligns with the classical realization of the merge sort algorithm, where both functions are typically implemented together.

emrebiyik commented:
Suggestion:
	- Combine merge and merge_sort in a Single File:
	•	These two functions are inherently interdependent. The merge function is a core helper function that works in tandem with merge_sort.
	•	Combining them into one file will improve maintainability and readability, as developers can see the full context of how these two functions interact without needing to navigate across multiple files.
	•	It aligns with the classical realization of the merge sort algorithm, where both functions are typically implemented together.

oleksandr-maksymikhin committed:
Added TypeVar for specifying input data types (int, float, str, bool)

Corrected code after code review. Added TypeVar for specifying input data types (int, float, str, bool). Small corrections by texting/comments

emrebiyik reviewed (approved):
Great work on the updates! The improvements to the docstring and the usage examples in the module-level documentation significantly enhance the clarity and usability of the code. 

Thank you for addressing the feedback effectively. Everything looks good to me, and I’m happy to approve this PR!
